### PR TITLE
Moved relu, conv2d, dropout, interpolate, cross_entropy, batch_norm and max_pool2d to C++ 

### PR DESCRIPTION
**Pull request by izdeby**
### Description
Moving to C++ these nn functions: 
- relu
- conv2d
- dropout
- interpolate
- cross_entropy
- batch_norm
- max_pool2d

### Testing:
1. Unit tests
2. Segmentation pipeline from pytorch/vision, master vs nt updated
**command**: _clear && python references/segmentation/train.py -j 0 -b 2 --device cuda --dataset coco --pretrained --test-only_ 

**master**:
Test:  [4700/5000]  eta: 0:00:46    time: 0.1620  data: 0.0397  max mem: 725
Test:  [4800/5000]  eta: 0:00:31    time: 0.1538  data: 0.0336  max mem: 725
Test:  [4900/5000]  eta: 0:00:15    time: 0.1595  data: 0.0383  max mem: 725
Test: Total time: 0:12:55
global correct: 91.9
mean IoU: 63.7

**nt**
Test:  [4700/5000]  eta: 0:00:56    time: 0.1947  data: 0.0392  max mem: 862
Test:  [4800/5000]  eta: 0:00:37    time: 0.1861  data: 0.0333  max mem: 862
Test:  [4900/5000]  eta: 0:00:18    time: 0.1894  data: 0.0357  max mem: 862
Test: Total time: 0:15:45
global correct: 91.9
mean IoU: 63.7

### Benchmark results:
fn _relu_tensor    avg(us): 145582.2344 std(us):  1661.2356 num_runs: 35
fn _relu_nt        avg(us): 153378.7812 std(us):   763.1132 num_runs: 33

fn _conv2d_tensor  avg(us): 185856.1875 std(us): 13552.3857 num_runs: 27
fn _conv2d_nt      avg(us): 193045.9375 std(us): 10189.9268 num_runs: 26

fn _batch_norm_tensor avg(us): 154030.5938 std(us):  9695.1162 num_runs: 33
fn _batch_norm_nt  avg(us): 153842.0156 std(us):  7826.2144 num_runs: 33

fn _max_pool2d_tensor avg(us): 72471.0547 std(us):  4896.7163 num_runs: 69
fn _max_pool2d_nt  avg(us): 71845.0547 std(us):   796.6418 num_runs: 70

fn _cross_entropy_tensor avg(us): 1031677.0000 std(us):  1064.3031 num_runs: 5
fn _cross_entropy_nt avg(us): 1024477.6875 std(us):   884.7659 num_runs: 5

fn _dropout_tensor avg(us): 562109.7500 std(us):   137.6644 num_runs: 9
fn _dropout_nt     avg(us): 562909.1250 std(us):  1605.0973 num_runs: 9

fn _interpolate_tensor avg(us): 489903.5000 std(us):   825.9186 num_runs: 11
fn _interpolate_nt avg(us): 405170.5312 std(us): 16439.2676 num_runs: 13

### TODO: 
1. fix max_pool2d overloads
2. cover with tests the optional parameters 
3. add input data validation
4. come up with better errors once NT is passed as an input. pybind 'eat' real error messages.

### CODE REVIEW

File: benchmarks/segmentation_layers.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
+import torch
+import nestedtensor
+import utils
+
+#
+# relu
+#
+def relu_tensor():
+    inputs = [
+            torch.randn(30, 500, 600),
```
Comment (cpuhrsch) [inline benchmarks/segmentation_layers.py:10]: This can be significantly shortened using list expressions.

```inputs = [torch.randn(30, 500, 600) for _ in range(6)]```

File: benchmarks/segmentation_layers.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    
+    a = torch.stack(inputs)
+
+    def _relu_tensor():
+        torch.nn.functional.relu(a)
+    return _relu_tensor
+
+def relu_nt():
+    inputs = [
+            torch.randn(30, 500, 600),
```
Comment (cpuhrsch) [inline benchmarks/segmentation_layers.py:26]: This sizes are "tensor compliant", so the benchmark should include something in its name that implies this. For now you're really only testing NestedTensor overhead with this.

File: nestedtensor/csrc/functions.cpp [hunk 3]
```diff
@@ -60,5 +63,178 @@ NestedTensor squeeze(
...
+                       const at::Tensor weight, 
+                       c10::optional<at::Tensor> bias, 
+                       c10::optional<std::vector<int64_t>> stride,
+                       c10::optional<std::vector<int64_t>> padding,
+                       c10::optional<std::vector<int64_t>> dilation,
+                       c10::optional<int64_t> groups) {
+  NestedTensor nt = input.data().contiguous();
+  TensorNode structure = nt.get_structure();
+  TensorNode res = map([&](at::Tensor t){
+      auto options = F::Conv2dFuncOptions().stride(stride.value())
```
Comment (cpuhrsch) [inline nestedtensor/csrc/functions.cpp:92]: You could move this outside and reference it from within the lambda.

File: nestedtensor/csrc/functions.cpp [hunk 4]
```diff
@@ -60,5 +63,178 @@ NestedTensor squeeze(
...
+                                           .padding(padding.value())
+                                           .dilation(dilation.value())
+                                           .groups(groups.value());
+      if (bias.has_value()) {
+          return F::conv2d(t.unsqueeze(0), weight, options.bias(bias.value())).squeeze(0);
```
Comment (cpuhrsch) [inline nestedtensor/csrc/functions.cpp:97]: Same with the bias.

File: nestedtensor/csrc/functions.cpp [hunk 5]
```diff
@@ -60,5 +63,178 @@ NestedTensor squeeze(
...
+                              const THPNestedTensor target,
+                              c10::optional<at::Tensor> weight,
+                              c10::optional<bool> size_average, // TODO: use
+                              c10::optional<int64_t> ignore_index,
+                              c10::optional<bool> reduce, // TODO: use
+                              c10::optional<std::string> reduction) {
+  TensorNode input_structure = input.data().contiguous().get_structure();
+  TensorNode target_structure = target.data().contiguous().get_structure();
+  F::CrossEntropyFuncOptions::reduction_t redct;
+  if (reduction.value() == "mean" || reduction.value() == "none") {
```
Comment (cpuhrsch) [inline nestedtensor/csrc/functions.cpp:159]: If don't you technically *have* to do this, but it surely yields a nicer error.

File: nestedtensor/csrc/python_functions.cpp [hunk 6]
```diff
@@ -49,5 +52,88 @@ void add_functions(
       },
       py::arg("dim") = nullptr);
 }
+
+void add_relu(pybind11::module m,
```
Comment (cpuhrsch) [inline nestedtensor/csrc/python_functions.cpp:56]: I'd just fold this into ```add_functions``` on line 14.

File: nestedtensor/nested/codegen/extension.py [hunk 7]
```diff
@@ -8,6 +8,17 @@
 # to efficiently implement a pointwise unary function such as cos
 # and then generalize it using the list of unary functions.
 
+def get_nn_C_functions():
```
Comment (cpuhrsch) [inline nestedtensor/nested/codegen/extension.py:11]: Just double check existing overwrites were removed.

File: nestedtensor/nested/monkey_patch.py [hunk 8]
```diff
@@ -134,6 +134,14 @@ def new_fn(self):
         setattr(NestedTensor, function_name + "_", _gen_fn(function_name + "_"))
         C_functions[getattr(torch, function_name)] = function_name
     # <
+    
+    # > NN functions supported by _C
+    for function_name in codegen.get_nn_C_functions():
+        setattr(NestedTensor,
```
Comment (cpuhrsch) [inline nestedtensor/nested/monkey_patch.py:140]: There is no "NestedTensor.conv2d" in pytorch etc.

I don't think any of these are methods outside of maybe relu.

File: nestedtensor/nested/codegen/extension.py (hunk 7 above)
Comment (cpuhrsch) [reply nestedtensor/nested/codegen/extension.py:11]: Also comment them out from other codegen lists if there are any

File: nestedtensor/nested/nested.py [hunk 9]
```diff
@@ -213,11 +213,104 @@ def nested_stride(self, dim=None):
     def __torch_function__(self, func, types, args=(), kwargs=None):
         _local_func = None
         if func in NestedTensor.__C_functions:
+            if 'interpolate' in str(func):
+                # size can be None, torch.Size, int or tuple of int
```
Comment (cpuhrsch) [inline nestedtensor/nested/nested.py:217]: None of this should be necessary.

File: nestedtensor/nested/nested.py [hunk 10]
```diff
@@ -213,11 +213,104 @@ def nested_stride(self, dim=None):
...
+                if kwargs['size'] is None:
+                    size = kwargs['size']
+                elif isinstance(kwargs['size'], tuple) or isinstance(kwargs['size'], torch.Size):
```
Comment (cpuhrsch) [inline nestedtensor/nested/nested.py:220]: You can enable overloads by registering multiple pybind functions under the same number but with different arguments. pybind infers the signatures. overload resolution is done in order of registration.

File: nestedtensor/nested/nested.py [hunk 11]
```diff
@@ -213,11 +213,104 @@ def nested_stride(self, dim=None):
...
+                                                                                               kwargs['weight'],
+                                                                                               kwargs['bias'],
+                                                                                               kwargs['training'],
+                                                                                               kwargs['momentum'],
+                                                                                               kwargs['eps']))
+
+            if 'max_pool2d' in str(func) or 'boolean_dispatch.<locals>' in str(func):
+                kernel_size = (args[1], args[1]) if type(args[1]) is not tuple else args[1]
+                stride = (kwargs['stride'], kwargs['stride']) if type(kwargs['stride']) is not tuple else kwargs['stride']
+                padding = (kwargs['padding'], kwargs['padding']) if type(kwargs['padding']) is not tuple else kwargs['padding']
```
Comment (cpuhrsch) [inline nestedtensor/nested/nested.py:254]: This kind of upcasting can also be done with overloading.

File: test/test_nested_tensor_functional.py [hunk 12]
```diff
@@ -24,78 +24,123 @@ def test_addmm(self):
...
 
-    def test_conv2d(self):
+    def test_nn_conv2d(self):
+        inputs = [
+            torch.randn(3, 500, 600),
+            torch.randn(3, 128, 128)
+        ]
+
+        # most of optional params
+        conv2d = torch.nn.Conv2d(3, 33, kernel_size=(3, 5), stride=(2, 1), padding=(4, 2), padding_mode='zeros', dilation=(3, 1), groups=1, bias=True)
```
Comment (cpuhrsch) [inline test/test_nested_tensor_functional.py:34]: Don't exercise the module but the functional. The module might start using the functional in unexpected ways.

File: test/test_nested_tensor_functional.py [hunk 13]
```diff
@@ -24,78 +24,123 @@ def test_addmm(self):
             [torch.rand(1, 4), torch.rand(1, 4), torch.rand(4, 4)]
         )
 
-    def test_conv2d(self):
+    def test_nn_conv2d(self):
```
Comment (cpuhrsch) [inline test/test_nested_tensor_functional.py:27]: In general attempt to exercise all overloads.

File: nestedtensor/nested/nested.py (hunk 9 above)
Comment (izdeby) [reply nestedtensor/nested/nested.py:217]: We discussed this and decided to cover all the cases. Seg pipeline calls with 2 different inputs. Also, the work is already done.

File: test/test_nested_tensor_functional.py (hunk 12 above)
Comment (izdeby) [reply test/test_nested_tensor_functional.py:34]: add module tests

**Comment by izdeby** [comment]
@cpuhrsch, should be good to go. 
Resolved most of the comments. Moved binding code and cleaned up C++ api to work with NestedTensor objects. Added tests for layers to cover torch.nn.X and torch.nn.functional.X where it applies. Some refactoring and cleaning up.

File: nestedtensor/csrc/functions.cpp [hunk 14]
```diff
@@ -60,5 +63,195 @@ NestedTensor squeeze(
...
+  }
+
+  TensorNode res = map([&, options](at::Tensor t){
+      return F::conv2d(t.unsqueeze(0), weight, options).squeeze(0);
+  }, structure);
+
+  return NestedTensor(std::move(res));
+}
+
+NestedTensor maxPool2d(NestedTensor input,
```
Comment (cpuhrsch) [inline nestedtensor/csrc/functions.cpp:116]: The name used in aten is [max_pool2d](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L1814). Think of these as overloads of native functions.

File: nestedtensor/nested/nested.py [hunk 15]
```diff
@@ -213,11 +213,63 @@ def nested_stride(self, dim=None):
     def __torch_function__(self, func, types, args=(), kwargs=None):
         _local_func = None
         if func in NestedTensor.__C_functions:
+            impl_args = [a._impl if isinstance(a, NestedTensor) else a for a in args]
```
Comment (cpuhrsch) [inline nestedtensor/nested/nested.py:216]: None of these changes should be required. You shouldn't have to change anything in nested.py.

File: nestedtensor/nested/nested.py [hunk 16]
```diff
@@ -213,11 +213,63 @@ def nested_stride(self, dim=None):
...
+                return NestedTensor(getattr(nestedtensor._C, NestedTensor.__C_functions[func])(*impl_args))
+
             assert len(args) == 1
             if kwargs is None:
                 return NestedTensor(getattr(nestedtensor._C, NestedTensor.__C_functions[func])(args[0]._impl))
-            assert len(kwargs) == 1 and 'out' in kwargs
-            return NestedTensor(getattr(nestedtensor._C, NestedTensor.__C_functions[func])(args[0]._impl, kwargs['out']._impl))
+            #assert len(kwargs) == 1 and 'out' in kwargs
+            if 'out' in kwargs:
+                return NestedTensor(getattr(nestedtensor._C, NestedTensor.__C_functions[func])(args[0]._impl, kwargs['out']._impl))
```
Comment (cpuhrsch) [inline nestedtensor/nested/nested.py:270]: You can still use the ```*args``` notation, in this case it would be ```*impl_args```.
